src/diffusers/image_processor.py

```diff
@@ -0,0 +1,160 @@
+        vae_scale_factor (`int`, *optional*, defaults to `8`):
+            scale factor in VAE, if do_resize is True, the image will be automatically resized to multipls of vae_scale_factor
+        resample (`str`, *optional*, defaults to `lanczos`):
+            Resampling filter to use if resizing the image.
+        do_normalize (`bool`, *optional*, defaults to `True`):
+            Whether to normalize the image to [-1,1]
+    """
+    
+    config_name = CONFIG_NAME
+
+    @register_to_config
+    def __init__(
+        self,
+        do_resize: bool = True,
+        vae_scale_factor: int = 8,
+        resample: str = "lanczos",
+        do_normalize: bool = True,
+        ):
+        super().__init__()
+
+    @staticmethod
+    def numpy_to_pil(images):
+        """
+        Convert a numpy image or a batch of images to a PIL image.
+        """
+        if images.ndim == 3:
+            images = images[None, ...]
+        images = (images * 255).round().astype("uint8")
+        if images.shape[-1] == 1:
+            # special case for grayscale (single channel) images
+            pil_images = [Image.fromarray(image.squeeze(), mode="L") for image in images]
+        else:
+            pil_images = [Image.fromarray(image) for image in images]
+
+        return pil_images
+    
+    @staticmethod
+    def numpy_to_pt(images):
+        """
+        Convert a numpy image  to a pytorch tensor
+        """
+        images = torch.from_numpy(images.transpose(0, 3, 1, 2))
+        return images
+
+    @staticmethod
+    def pt_to_numpy(images):
+        """
+        Convert a numpy image  to a pytorch tensor
+        """
+        images = images.cpu().numpy().transpose(0, 2, 3, 1)
+        return images
+
+    @staticmethod
+    def normalize(images):
+        """
+        Normalize an image array to [-1,1]
+        """
+        return 2.0 * images - 1.0
+    
+    def resize(self, images: PIL.Image.Image) -> PIL.Image.Image:
+        """
+        Resize an PIL image. Both height and width will be resized to integer multiple of vae_scale_factor
+        """
+        w, h = images.size
+        w, h = map(lambda x: x - x % self.vae_scale_factor, (w, h))  # resize to integer multiple of vae_scale_factor
+        images = images.resize((w, h), resample=PIL_INTERPOLATION[self.resample])
+        return images
+
+    def encode(
+        self, 
+        image: Union[torch.FloatTensor, PIL.Image.Image, np.ndarray],
+    ) -> torch.Tensor:
+
+        """
+        Preprocess the image input, accpet formats in PIL images, numpy arrays or pytorch tensors"
+        """
+        # convert PIL or list of PIL into numpy 
+        if isinstance(image, PIL.Image.Image):
+            image = [image]
+
+        if isinstance(image[0], PIL.Image.Image):
+            if self.do_resize:
+                image = [self.resize(i) for i in image]
+            image = [np.array(i).astype(np.float32) / 255.0 for i in image]
```

Comment:
```suggestion
            image = [np.array(i).astype(np.float32) / 255.0 for i in image]
        elif self.do_resize:
            # Currently we only support resizing for PIL so in case np or torch is used AND resizing is activating (which it is by default) then let's do the following: - 1. Check if the image sizes are not a multiple of `self.vae_scale_factor` => If it's not the case we throw a nice error
```

Reply:
This `elif` statement is not implemented yet, please make sure to add such a quick test + error message.

Reply:
add the test + error message here on Line 133
https://github.com/huggingface/diffusers/blob/image-processor/src/diffusers/image_processor.py#L133

because it is easier to find out the width and height of the image after converting the into torch tensors with 4 dimensions, otherwise it's a little bit complex (i.e. could be 3d or 4d, torch tensors are channel-first and numpy arrays are channel-last)

let me know if this is ok

Reply:
No I think there has been a misunderstanding here. The code as is currently is not very readable because:
- If the input is PIL and is of a weird shape -> then we resize it to a correct shape and the error message below will never be hit
- If the input is torch and numpy and of a weird, then in this case we throw an error message that is not great because it just tells the user that the shapes are incorrect, but not that the user could have also changed the input to PIL and it would have been resized correctly.

In addition this is not readable - we want error messages close to where they are relevant. Let's try to change that here so that the error message says that either the shapes have to changed or the user should pass a PIL input + let's move the error message to the part of the code it corresponds to not over complicate the logic here.